# Perform 3

In this assignment, you will demonstrate your mastery with wrangling data through different formats, with missing data, and working with text.

You can either print answers directly from your code or write them in the markdown cells below your code. Either way, make sure that your answers are visible and can be easily read in the final notebook you turn in.

## Part 1: Text wrangling and regular expressions
In this part we will work with the citation file exported from the [Nature Review Article](https://www.nature.com/articles/s41586-020-2649-2) *Array Programming with NumPy*. Below we read the file into the Python variable `cite` and print the result for you to preview.

In [1]:
# Run but do not modify this code
with open("numpy_nature.txt") as f:
    cite = f.read()

print(cite)

### Question 1 (4 points)
There are several authors, each recorded on a separate line beginning with `AU`. Create a Python list of all of the author names formatted as in the file but without the extra characters and whitespace (i.e., without the `AU  - ` or the newline `\n` characters). Your list should be of the form `['Harris, Charles R.', 'Millman, K. Jarrod', ..., 'Oliphant, Travis E.']`. When you are finished, print the resulting list. 

In [2]:
newtxt=cite.split("-")

newtxt.remove(newtxt[0])
newtxt.remove(newtxt[0])
newtxt=[x.strip("\nAU,\nPY, ,") for x in newtxt]
index = newtxt.index('Oliphant, Travis E.')
print(newtxt[0:index+1])

### Question 2 (4 points)
Create a Pandas DataFrame that contains three columns: one for first names, one for middle names, and one for last names for all of the authors. You should use descriptive column names, but you can use the default primary index (the row labels) of 0, 1, 2, etc. Thus, the first few rows of your dataframe might look like the table pictured below. You are welcome to use the results of the prior question to asnwer this problem.

|      | first      | middle     | last         |
| ---- | ---------- | ---------- | ------------ |
| 0	   | Charles    | R.         | Harris       |
| 1	   | K.	        | Jarrod     | Millman      |
| 2	   | Stéfan     | J.         | van der Walt |
| 3    | Ralf       |            | Gommers      |
| 4	   | Pauli      |            | Virtanen     |   

Note that some authors do not have any middle names, in which case you can leave the middle name column blank. When you are finished, display the first 10 rows of the resulting dataframe by calling `pd.head(10)` (where Pandas has been imported as `pd`).

In [3]:
# Run but do not modify this code
import pandas as pd

In [4]:
namedftxt=newtxt[0:index+1]
first=[]
middle=[]
last=[]
FandM=[]
for x in namedftxt:
    sepname=x.split(",")
    last.append(sepname[0])
    if(len(sepname)>1):
        FandM.append(sepname[1].lstrip())
    else:
        FandM.append(sepname[0].lstrip())

for i in FandM:
    sepname2=i.split(" ")
    first.append(sepname2[0])
    if(len(sepname2)>1):
        middle.append(sepname2[1])
    else:
        middle.append(" ")
cols=tuple(zip(first, middle,last))

In [5]:
df = pd.DataFrame(cols, columns =['first', 'middle', 'last'], dtype = str) 


In [6]:
df.head(10)

first  middle          last
0    Charles      R.        Harris
1         K.  Jarrod       Millman
2     Stéfan      J.  van der Walt
3       Ralf               Gommers
4      Pauli              Virtanen
5      David            Cournapeau
6       Eric                Wieser
7     Julian                Taylor
8  Sebastian                  Berg
9  Nathaniel      J.         Smith

### Question 3 (6 points)
Below we extract the abstract from the citation and store it in a string variable `abstract`. Write regular expressions to answer the following questions about the abstract.

1. Print the starting index of everywhere `NumPy` appears in the abstract (i.e., the index of the `N` wherever `NumPy` occurs in the `abstract` string).
2. Print all of the capitalized words in `abstract`, including words with extra capitalized letters like `NumPy`.
3. Print all of the words that immediately follow `NumPy`, but do not include the word `NumPy` itself. Note that in one occurrence it is hyphenated `NumPy-like`, in which case your code can return `-like` or `like` as you prefer.

In [7]:
# Run but do not modify this code
import re
abstract_query = re.compile(r"AB  - (.+)")
abstract = re.search(abstract_query, cite).group(1)
print(abstract)

Array programming provides a powerful, compact and expressive syntax for accessing, manipulating and operating on data in vectors, matrices and higher-dimensional arrays. NumPy is the primary array programming library for the Python language. It has an essential role in research analysis pipelines in fields as diverse as physics, chemistry, astronomy, geoscience, biology, psychology, materials science, engineering, finance and economics. For example, in astronomy, NumPy was an important part of the software stack used in the discovery of gravitational waves1 and in the first imaging of a black hole2. Here we review how a few fundamental array concepts lead to a simple and powerful programming paradigm for organizing, exploring and analysing scientific data. NumPy is the foundation upon which the scientific Python ecosystem is constructed. It is so pervasive that several projects, targeting audiences with specialized needs, have developed their own NumPy-like interfaces and array object

In [8]:
#1
arry=re.finditer(r'NumPy', abstract)
for i in arry:
    print(i.start())
#2
query1 = re.compile(r"[A-Z]+[A-Z,a-z,-]*")
print(re.findall(query1, abstract))
print("")
#3
query2 = re.compile(r'NumPy([" ",-]+[A-Z,a-z,-]+)')
print(re.findall(query2, abstract))

171
469
768
962
1051
['Array', 'NumPy', 'Python', 'It', 'For', 'NumPy', 'Here', 'NumPy', 'Python', 'It', 'NumPy-like', 'Owing', 'NumPy', 'API']

[' is', ' was', ' is', '-like', ' increasingly']


## Part 2: Cleaning up more system logs CSV
In this part we work with a piece of messy tabular data in the form of a poorly formatted csv file containing data about programs running on computer systems. It contains all of the data about system time and memory from Practice 3, but also includes new information about user ids and machine ids, and some data are missing in every column. (The user ids are made up and do not correspond to any real individuals).  

### Question 4 (6 points)
Below, we import the dataset using the Pandas `read_csv` function that creates a dataframe. Run the code; it will preview the first five rows. 

In [9]:
# Run but do not modify this code
import pandas as pd
import numpy as np
sys_df = pd.read_csv("more_monitor.csv")
sys_df.head()

System Time second   System Memory GB System Memory MB System User ID  \
0                       ?                  ?                ?  User ID: yw22   
1  System Time: 40 second   System Mem: 3 Gb           382 Mb              ?   
2                       ?   System Mem: 2 Gb           271 Mb   User ID: tp7   
3  System Time: 31 second   System Mem: 3 Gb           493 Mb              ?   
4  System Time: 37 second   System Mem: 3 Gb           411 Mb              ?   

       System Machine ID  
0     Machine ID: Carrot  
1                      ?  
2  Machine ID: Asparagus  
3   Machine ID: Eggplant  
4  Machine ID: Asparagus

There are several formatting issues with the default import. Address the following.

1. The `System User ID` and `System Machine ID` contain String data with the redundant information `User ID: ` and `Machine ID: ` in every row that has data. Remove these prefixes so that the columns only contain the user ids and machine ids themselves (for example, the first row should just have `yw22` in the `System User ID` column and `Carrot` in the `System Machine ID` column. 
2. The first three columns for `System Time second`, `System Memory GB` and `System Memory MB` contain numerical data but are currently formatted as strings with redundant prefix information repeating the column label and missing data represented as the string `?` instead of the Numpy `NaN` sentinel value. Fix this so that each value in the first three columns is either a single numerical value or `NaN` (note, you should use the actual `np.NaN` sentinal value, not just the String with the characters `N`, `a`, and `N`). For example, when you are done, the first three columns of the first row should all have `NaN` values, the second row should be `40`, `3`, and `382`, and so on. Note that the rows at index `400` and on have System Time recorded in minutes instead of seconds, be sure to convert these to seconds by mulitplying by 60.
3. Currently the System Memory is split accross two columns, one for the GB and one for the MB. For example, the total memory of the first program is 3 GB and 414 MB. Instead, represent the full system memory in the `System Memory GB` column, and get rid of the `System Memory MB` column. To do so, you need to convert the values in the MB column to GB (1 MB is 0.001 GB) and add that to the GB column, then use the [`drop` method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html). Missing values should remain missing after this transformation.

When you are finished, `sys_df` should have the above issues corrected. Run both of the cells with `sys_df.head()` and `sys_df.tail()` to show the first and last few rows of your dataframe.

In [10]:
#1
sys_df['System User ID']=sys_df['System User ID'].str.lstrip('User ID:')

sys_df['System Machine ID']=sys_df['System Machine ID'].str.lstrip('Machine ID:')
#2
sys_df['System Time second']=sys_df['System Time second'].str.extract(r'([0-9]+)')
sys_df['System Time second']=sys_df['System Time second'].astype(float)
arr=sys_df['System Time second']
arr1=arr[:400]
arr2=arr[400:]*60
sys_df['System Time second']=np.concatenate((arr1, arr2))
sys_df['System Memory GB']=sys_df['System Memory GB'].str.extract(r'([0-9]+)').astype(float)
sys_df['System Memory MB']=sys_df['System Memory MB'].str.extract(r'([0-9]+)').astype(float)
#3
sys_df['System Memory GB']=sys_df['System Memory GB']+(sys_df['System Memory MB']/1000)
sys_df=sys_df.drop(['System Memory MB'], axis=1)




In [50]:
# Run but do not modify this code
sys_df.head()

System Time second  System Memory GB System User ID System Machine ID
0                 NaN               NaN           yw22            Carrot
1                40.0             3.382              ?                 ?
2                 NaN             2.271            tp7         Asparagus
3                31.0             3.493              ?          Eggplant
4                37.0             3.411              ?         Asparagus

In [12]:
# Run but do not modify this code
sys_df.tail()

System Time second  System Memory GB System User ID System Machine ID
995              1860.0             2.258            zm3         Asparagus
996              1860.0             3.403            bk4                 ?
997              2160.0             3.350           yw22            Carrot
998              2640.0             3.366           yw22         Asparagus
999              1200.0             3.490            bk4         Asparagus

### Question 5 (6 points)
The `sys_df` dataframe from question 3 should now be a little easier to read and use. Answer the following questions about `sys_df`.

1. How many rows are missing data in the `System Machine ID` column?
2. What is the average value of `System Memory GB` among the rows that are missing data in the `System User ID` column? 
3. How many rows are missing data in both the `System Time second` and `System Memory GB` columns?

Note: It is not necessary to complete all of question 4 in order to answer some of these questions, and we will also look at your code for partial credit.

In [49]:
#1
print(sys_df['System Machine ID'].str.count(r'([?])').sum())
#2
boolarr=sys_df['System Machine ID'].str.contains('[?]')
regarr=sys_df['System Memory GB']
print(np.nansum(regarr[boolarr]))
#3
boolarr1=sys_df[(np.isnan(sys_df['System Time second'])) & (np.isnan(sys_df['System Memory GB']))]
print(len(boolarr1))



196
403.06
69


### Answer 5
Write your answer here if not printed above 

## Part 3: Wrangling FDA JSON Dataset 
In this part we work with a messy JSON dataset containing information about several drugs labels.

### Question 6 (6 points)
Below we import the `FDADrugLabel.json` file into the `labels` variable. This is the same dataset as the practice. The resulting Python object is somewhat messy; we encourage you to explore the data before answering the questions.

In [151]:
# Run but do not modify this code
import json
with open("FDADrugLabel.json") as f:
    labels = json.load(f)


Answer the following questions.

1. Print the average number of key/value (or name/value) pairs for the drugs.
2. Print the list of all of the `manufacturer_names` without any other information. `manufacturer_names` are not a top level key/name, you will need to search for where they are located and how to extract them.
3. Print how many drugs contain the string `child` anywhere in their `warnings` (including as part of larger strings like `children`). `warnings` is a top level key/name, but some drugs are missing this data and the information is contained in strings within lists of length 1. 

In [150]:
#1
totalsum=0
for i in range(0,len(labels)):
    totalsum=totalsum+len(labels[0].values())
print(totalsum/len(labels)/len(labels[0].values()))
#2
for i in range(0,len(labels)):
    trying=labels[i].get("openfda")
    print(trying.get('manufacturer_name'))
#3
counter=0
for i in range(0,len(labels)):
    warnings=labels[i].get("warnings"," ")
    queries=re.compile(r'([Cc]hild)')
    if(len(re.findall(queries,warnings[0]))>0):
        counter=counter+1
print(counter)


1.0
Nature and Health Beauty Co., Ltd.
Silver Star Brands
Johnson & Johnson Consumer Inc.
Proficient Rx LP
Energique, Inc.
Energique, Inc.
Amerisource Bergen
Seroyal USA
Proficient Rx LP
MODECOS CO., LTD.
United Exchange Corp.
King Bio Inc.
Aurobindo Pharma Limited
Energique, Inc.
SMART SENSE (Kmart)
Mentor Lab
BCM Ltd
Pearl World INC.
Dolgencorp, Inc. (DOLLAR GENERAL & REXALL)
Medline Industries, Inc.
13


### Answer 6
Write your answer here if not printed above